### ISSUE TITLE
SRAM Memory needed just for inference

### ISSUE THREAD

**Issue by Deepakpy** [Fraunhofer-IMS/AIfES_for_Arduino#26]
Hallo I would like to do training on one MCU which has a bit larger sram and want to use the updated model parameters on a smaller MCU with less memory.

I see to do the training i need a larger scratchpad memory but if i just want to do inference and no training on this smaller mcu. Can i know if it is possible and would it be still possible to do inference by reducing the memory size on sram.

Thanks for your inputs in advance.

**Comment by Johannes-IMS** [comment]
Hello @Deepakpy,

Thanks for trying out AIfES.

You need less memory when you only want to perform the inference, as gradients, for example, do not need to be stored. So you can train your model on an MCU with bigger SRAM and then run the inference on an MCU with smaller SRAM.
Of course, you need to share the parameters between these two MCUs (more details on [parameter memory](https://fraunhofer-ims.github.io/AIfES_for_Arduino/_tutorial_training_f32.html#autotoc_md21)).

More details regarding memory allocation for training and inference (without parameter memory, which needs to be supplied separately) can be found in the documentation ([Allocating memory for inference](https://fraunhofer-ims.github.io/AIfES_for_Arduino/_tutorial_inference_f32.html#autotoc_md14); [Allocating memory for training](https://fraunhofer-ims.github.io/AIfES_for_Arduino/_tutorial_training_f32.html#autotoc_md27)). 

cc @LarsWulfert

**Comment by Deepakpy** [comment]
Thanks for your inputs. I used the [keras2aifes.convert_to_fnn_f32](https://fraunhofer-ims.github.io/AIfES-Converter/advanced_usage.html) to convert a simple mnist detection mlp model. This generated weights.h file (size:292 kB) and a model_f32_fnn.h file(size:4.37 kB).

I have two questions on this and related to previous reply from you:
1. Is the minimum sRAM needed is equal to or more than the size of weights.h header file(i assume this is the parameter memory needed)? 
2. Can i just use [Allocate memory for inference](https://fraunhofer-ims.github.io/AIfES_for_Arduino/_tutorial_inference_f32.html#autotoc_md14)  by loading the model from weights.h header file as flat_weights_array  and only set parameter_memory for just inference. From what i understood the parameter_memory is large compared to the inference memory is their a way to minimise it?

**Comment by Johannes-IMS** [comment]
Thanks for getting back with some more details. 

Regarding your two questions:

1. The needed size of SRAM depends on different aspects. One of these aspects is your model architecture. Moreover, if you want to train the model, you will need even more storage space compared to inference only. As you already used the AIfES-Converter, you should have noticed that the parameters (in weights.h file) are declared `const`. Therefore, these parameters are most often automatically stored in flash and not in SRAM. Consequently, the SRAM is substantially smaller for inference than the parameter size. However, if you want to perform training, these parameters need to be saved also in SRAM, as these parameters get updated during training.
2. This is already implemented in the model_f32_fnn.h file in the same way as you described it. Only the inference memory should be allocated on SRAM, and the parameter memory should be located in flash (by declaring the parameters as `const`). Additionally, you can use advanced storage management solutions such as quantization to optimize the size of your ANN.

**Comment by Deepakpy** [comment]
Okay it's clear now the parameters memory is assigned to the flash but now when i have the new parameters weights from the bigger sRAM mcu where traning happens. How would this weights be loaded to the smaller sRAM mcu? I assume if i use any buffer to write new weights to the smaller sRAM MCU than i need to have to store the new weights to the sRAM ?
Can you please let me know what all might be the possible options and which would use minimal RAM?

**Comment by Johannes-IMS** [comment]
Hello @Deepakpy,

I understand that you need to transfer the parameter memory between your two MCUs. While the specifics of this process fall outside the scope of AIfES, I would like to offer some guidance. One potential solution could be to stream data from your larger MCU to the smaller one, provided that the smaller MCU supports writing your received parameters into its Flash memory.

If you require assistance with memory management, which is beyond AIfES's scope, we would be more than happy to collaborate with you to find a suitable solution. Please feel free to reach out to us through our [contact page](https://www.ims.fraunhofer.de/en/Business-Unit/Industry/Industrial-AI/Artificial-Intelligence-for-Embedded-Systems-AIfES/AIFES_Cooperation.html) by clicking on "Send email."

Additionally, as mentioned previously, you can utilize quantization to reduce the size of the parameter memory. You can find more details on this in our [documentation](https://fraunhofer-ims.github.io/AIfES_for_Arduino/_tutorial_inference_q7.html) and explore these [examples utilizing quantization](https://github.com/Fraunhofer-IMS/AIfES_for_Arduino/tree/main/examples/0_Universal/1_XOR_Q7).
